src/v/cluster/archival/segment_reupload.cc

```diff
@@ -1014,6 +1021,13 @@ ss::future<candidate_creation_result> segment_collector::make_upload_candidate(
               .end_offset = _end_inclusive,
               .reason = candidate_creation_error::upload_size_unchanged};
         }
+        if (!is_reupload_mode(_mode)) {
+            vlog(
+              archival_log.debug,
+              "New segment upload already appears in manifest: {}",
+              _segments.front());
+            co_return candidate_creation_error::concurrency_error;
```

Comment:
This is fine and avoids the problem at hand but I think for non-reupload mode we shouldn't even drop into the to_replace check.

The check is still useful as in case someone updates the manifest concurrently we don't try to do an useless upload. However, if the new leader didn't yet update the manifest, from what I'm reading, we would still do the useless upload which we won't be able to replicate anyway.

Instead, there should be some sort of abort source present somewhere which would abort the operations/in-progress upload when we give away the leadership.

wdyt?

Reply:
I don't disagree; sounds like a worthy optimization. I have a fair amount of archival code moving around on various branches, so I'll have to punt on that a little bit. I can put it on my TODO list though, certainly.